main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs

```diff
@@ -567,14 +591,182 @@ internal bool ProcessKeyEvent (Gdk.EventKey ev)
+
+		GtkNSViewHost GetGtkNSViewHostFromView (AppKit.NSView view) => view == null ? null : viewHosts.FirstOrDefault (s => s.Content == view);
+
+		static bool CanBecomeKeyView (AppKit.NSView view, IGtkViewHostContentView gtkViewHostContentView)
+		{
+			if (!view.CanBecomeKeyView)
+				return false;
+
+			if (view is AppKit.NSScrollView) {
+				return false;
+			}
+
+			if (view is AppKit.NSScroller) {
+				return false;
+			}
+
+			if (view is AppKit.NSClipView) {
+				return false;
+			}
+
+			if (gtkViewHostContentView != null && !gtkViewHostContentView.CanBecomeKeyViewFromGtkViewHost (view)) {
+				return false;
+			}
+			return true;
+		}
+
+		static bool AllowsRecursivityFocus (AppKit.NSView view)
+		{
+			if (view.Hidden) {
+				return true;
+			}
+			//this views are never included in the keyloop but they can be focused, we don't want to include it
+			if (view.Superview is AppKit.NSClipView clip && clip.Superview is AppKit.NSTextField) {
+				return true;
+			}
+			return false;
+		}
+
+		static void AddRecursivelyFocusableViews (AppKit.NSView view, List<AppKit.NSView> toAddViews, bool addViewBeforeChildren, AppKit.NSView contentView)
+		{
+			if (AllowsRecursivityFocus (view)) {
+				return;
+			}
+
+			if (addViewBeforeChildren && CanBecomeKeyView (view, contentView as IGtkViewHostContentView) && contentView != view)
+				toAddViews.Add (view);
+
+			if (view.IsFlipped) {
+				foreach (var subview in view.Subviews.OrderBy (s => s.Frame.Left).ThenBy (h => h.Frame.Top)) {
+					AddRecursivelyFocusableViews (subview, toAddViews, addViewBeforeChildren, contentView);
+				}
+			} else {
+				foreach (var subview in view.Subviews.OrderBy (s => s.Frame.Left).ThenByDescending (h => h.Frame.Top)) {
+					AddRecursivelyFocusableViews (subview, toAddViews, addViewBeforeChildren, contentView);
+				}
+			}
+
+			if (!addViewBeforeChildren && CanBecomeKeyView (view, contentView as IGtkViewHostContentView) && contentView != view)
+				toAddViews.Add (view);
+		}
+
+		//this method scales the hierarchy of a native view embedded in a GtkNSViewHost in search of the content view
+		AppKit.NSView GetGtkNSViewHostContentView (AppKit.NSView nSView)
+		{
+			if (nSView.Superview == null) {
+				//this could include cases where native views are in the toolbar
+				return null;
+			}
+			if (IsGtkQuartzView (nSView.Superview)) {
+				return nSView;
+			}
+			var parent = GetGtkNSViewHostContentView (nSView.Superview);
+			return parent;
+		}
+
+		const string GdkQuartzViewTypeName = "<GdkQuartzView";
+		static bool IsGtkQuartzView (AppKit.NSView view) => view.ToString ().StartsWith (GdkQuartzViewTypeName);
+
+		readonly List<Gtk.GtkNSViewHost> viewHosts = new List<Gtk.GtkNSViewHost> ();
+		internal void RegisterEmbededView (Gtk.GtkNSViewHost gtkNSViewHost)
+		{
+			if (gtkNSViewHost == null) {
+				throw new NullReferenceException ("cannot register a null GtkNSViewHost");
```

Comment:
```suggestion
				throw new ArgumentNullException ("cannot register a null GtkNSViewHost");
```

`NullReferenceException` is usually thrown by the runtime to signal null access, and we're not doing a null access here, we're signaling API mis-use.